rfcs/20250430-host-side-scalars/README.md

```diff
@@ -0,0 +1,180 @@
+# Host-side scalars Support in oneDNN Primitives
+
+## Motivation
+
+Currently, primitives assume that all inputs passed at execute time
+are on the same device as the one they execute on (with the exception
+of reorder).
+
+A new request is to pass scalar values that reside on host device
+to GPU kernels at execution time.
+A good example is the scale for SDPA operator in Pytorch [^1], which is
+a python variable on the host and not a torch.tensor object.
+
+There are two challenges here:
+- How to specify to primitive descriptor creation that a given input is
+  a host scalar?
+    - Currently, we don't have any engine specified to input memory descriptors.
+- How to pass host-side scalars to the execute function?
+  - Currently, we only take dnnl::memory objects.
+
+## How to specify an input as a host-side scalar value to primitive desc creation?
+
+### Option 1: do nothing and take host-side scalars as primitive descriptor parameters
+
+Here user would pass the host side scalar as a parameter to primitive
+creation, as is done today. The main caveats here are:
+- It requires user to recreate primitive descriptors and primitives
+  every time the host-side scalar changes value.
+- If the implementation jits a kernel that depends on that parameter,
+  we can get large overheads due to re-jitting.
+
+Pros:
+- oneDNN API can already handle this for parameters that do not
+frequently change.
+
+Cons:
+- Forces user to recreate primitive descriptor and primitive objects
+  every-time that host-side scalar value changes.
+- Potential extra jitting overhead if internal implementation considers that parameter constant.
+
+This option is not recommended if the host-side scalar value is
+expected to frequently change.
+
+### Option 2 (recommended): expose a new memory descriptor kind
+
+The idea here is to:
+- expose host-side scalars through a new memory descriptor create function
+- have user pass a memory with cpu engine to execute function.
+
+The new kind of memory descriptor would be created through a dedicated function
+It would be very lightweights and be limited:
+- only scalar value, so no `ndims`, no `dims`, ...
+- only scalar datatype can be configurable by user.
+
+```C
+dnnl_status_t DNNL_API dnnl_memory_desc_create_host_scalar(
+        dnnl_memory_desc_t *memory_desc, dnnl_data_type_t data_type);
+```
```

Comment:
Is there any implication or consideration about using this API and host-side scalar memory in a CPU only build and CPU primitives? It seems we don't a `host-side scalar` concept in that case. Do we expect that the CPU primitives will work with both a normal scalar memory or a host-side scalar memory (created from this new API)?

Reply:
Updated the recommended option and API to not have engine, POC link is available, so resolving this thread.
Feel free to reopen / start a different one if any clarification is needed still.